Projects/Shapes/Shapes/Shapes/DynamicArray.h

```diff
@@ -0,0 +1,341 @@
+
+	void Clear()
+	{
+		Size = 0;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void PushBack(const T& Item)
+	{
+		if (Size == Capacity)
+		{
+			if (Capacity == 0)
+			{
+				Reserve(1);
+			}
+			Reserve(Capacity * 2);
+		}
+		Array[Size] = Item;
+		Size++;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void PopBack()
+	{
+		if (Size > 0)
+		{
+			Size--;
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool Insert(const T& Item, const int Index)
+	{
+		if (Index > Size)
+		{
+			return false;
+		}
+
+		if (Index == Size)
+		{
+			PushBack(Item);
+			Size++;
+			return true;
+		}
+
+		if (Size == Capacity)
+		{
+			Reserve(Capacity * 2);
+		}
+		
+		for (int i = Size; i > Index; i--)
+		{
+			Array[i] = Array[i-1];
+		}
+
+		Array[Index] = Item;
+		Size++;
+		return true;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool Erase(const int Index)
+	{
+		if (Index > Size)
+		{
+			return false;
+		}
+
+		if (Index == Size)
+		{
+			PopBack();
+			return true;
+		}
+
+		for (int i = 0; i < Size - 1; i++)
+		{
+			Array[i] = Array[i + 1];
+		}
+		Size--;
+		return true;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Resize(const int NewSize)
+	{
+		if (Size > Capacity)
+		{
+			Reserve(NewSize);
+		}
+		Size = NewSize;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Resize(const int NewSize, const T& Item)
+	{
+		if (Size <= NewSize)
+		{
+			Size = NewSize;
+		}
+		else
+		{
+			if (NewSize > Capacity)
+			{
+				Reserve(NewSize);
+			}
+			for (int i = Size; i < NewSize; i++)
+			{
+				Array[i] = Item;
+			}
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Append(const TDynamicArray<T>& OtherArray)
+	{
+		Reserve(Size + OtherArray.GetSize());
+
+		for (int i = 0; i < OtherArray.GetSize(); i++)
+		{
+			PushBack(OtherArray[i]);
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------
+	//											CUSTOM METHODS
+	//-------------------------------------------------------------------------------------------
+
+	void RefactorCapacityTo(const int NewCapacity)
+	{
+		if (!Reserve(NewCapacity))
+		{
+			if (NewCapacity == Size)
+			{
+				return;
+			}
+
+			T* NewArray = new T[NewCapacity];
+			if (Size > NewCapacity)
+			{
+				Size = NewCapacity;
+			}
+
+			for (int i = 0; i < Size; i++)
+			{
+				NewArray[i] = Array[i];
+			}
+
+			Capacity = NewCapacity;
+			delete[] Array;
+			Array = NewArray;
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Fill(const T& Item, const bool FillToCapacity)
+	{
+		if (FillToCapacity)
```

Comment:
Nice, but this is the perfect scenario for a ternary operation. When your code is exactly the same except for the value of one variable, it may be better to simply do something like
```
const int FillLimit = bFillToCapacity ? Capacity : Size;
for (int i = 0; i < FillLimit; ++i)
```

That way you avoid long functions and code duplication